# SI 330: Data Manipulation 
## 17 - SQL

### Dr. Chris Teplovs, School of Information, University of Michigan
<small><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

# <FONT COLOR="red">NOTE: Please install the ```pymysql``` module (conda install pymysql)

# SQL Outline

## Day 1

* AWS Overview
  * EC2, RDS

* Tasks
  * log into AWS Console
  * create EC2 instance
    * yum install mysql
    * wget .sql
    
## Day 2

* RDS
* MySQL

* Tasks
  * create RDS instance
  * load SQL file from EC2
  * browse with MySQLWorkbench
  
## Day 3

* pymysql

* Tasks
  * query goodreads
  
## Day 4

* BigQuery, Homework


## SQL

In general, we are going to:
```
SELECT some_fields FROM some_table
WHERE some_condition_exists
[ GROUP BY some_field ]
[ ORDER BY some_field ]
```

For example:
```
SELECT authors FROM books;
```
**Note:** the convention is to use UPPERCASE for SQL reserved words, and lowercase for the specifics of your instance

If we want all the fields:
```
SELECT * FROM books;
```

Adding a condition:
```
SELECT * FROM books
WHERE original_publication_year >= 2008;
```

And you can also group by a field:
```
SELECT original_publication_year,COUNT(*) FROM books
GROUP BY original_publication_year;
```

## Introduction to pymysql



```pymysql``` is a python interface to SQL servers (including MySQL, sqlite, etc.).  There are other interfaces (such as SQLAlchemy), but pymysql is the most straight-forward to work with. (Note caveats that we'll discuss in class.)

In [2]:
!conda install -y pymysql

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\jodiy\Anaconda3.1

  added / updated specs:
    - pymysql


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.3.9           |           py36_0         156 KB
    conda-4.6.8                |           py36_0         1.7 MB
    openssl-1.1.1b             |       he774522_1         5.7 MB
    pymysql-0.9.3              |           py36_0          83 KB
    ------------------------------------------------------------
                                           Total:         7.7 MB

The following NEW packages will be INSTALLED:

  pymysql            pkgs/main/win-64::pymysql-0.9.3-py36_0

The following packages will be UPDATED:

  certifi                                 2018.11.29-py36_0 --> 2019.3.9-py36_0
  conda                                        4.6.4-py36_0 --> 4.6.8-py36_0
  ope

'C:\Users\jodiy\Anaconda3.1\condabin\.)' is not recognized as an internal or external command,
operable program or batch file.


In [3]:
import pymysql

Fill in the ```host```, ```user``` and ```password``` fields in the next code block.  The host is the RDS endpoint; user should probably be your uniqname, and password should be whatever you set you database password to.

Run the block to establish a connection to your RDS-based MySQL server:

In [4]:
connection = pymysql.connect(host='si330-joyip-instance.cc66ikpymjd4.us-east-1.rds.amazonaws.com', # remove this before submitting
                             user='', # remove this before submitting
                             password='', # remove this before submitting
                             db='goodreads',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

Here's a "pythonic" way to query your database.  THe following is a bit 

In [5]:
with connection.cursor() as cursor:
    cursor.execute("SELECT * FROM books WHERE original_title = 'The Hunger Games'")
    results = cursor.fetchall()
    for result in results:
        print(result['original_title'],result['authors'])

The Hunger Games Suzanne Collins


### COUNT

In [6]:
with connection.cursor() as cursor:
    cursor.execute("SELECT COUNT(*) FROM books WHERE authors = 'Suzanne Collins'")
    result = cursor.fetchone()
    print(result)

{'COUNT(*)': 9}


We can rename awkward column names by using the ```AS``` operator:

In [7]:
with connection.cursor() as cursor:
    cursor.execute("SELECT COUNT(*) AS cnt FROM books WHERE authors = 'Suzanne Collins'")
    result = cursor.fetchone()
    print(result)

{'cnt': 9}


### DISTINCT

Sometimes, we want to know the number of distinct (unique) values of a field.  Here are two ways of doing this:

In [8]:
with connection.cursor() as cursor:
    cursor.execute('SELECT DISTINCT authors FROM books')
    results = cursor.fetchall()
    print(len(results))

4669


In [9]:
with connection.cursor() as cursor:
    cursor.execute('SELECT COUNT(DISTINCT authors) FROM books')
    result = cursor.fetchone()
    print(result)

{'COUNT(DISTINCT authors)': 4669}


### <font color="magenta">Q1: Which of the above two ways of finding distinct values is more efficient?  Why?</font>

The second is more efficient because fetchone returns a single cell while fetchall returns 4669 rows.

## SQL from pandas

We are going to continue to use our connection object from above in this section.  Pandas plays nicely with pymysql.

In [10]:
import pandas as pd

Just like what we found with read_csv (a relief compared to CSVDictReader), we have a read_sql method that allows us to create a dataframe based on an SQL query:

In [11]:
df = pd.read_sql('SELECT * FROM books',con=connection)
len(df)

10000

In [15]:
df.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   
4        5               4671          4671   245494         1356  743273567   

         isbn13                       authors  original_publication_year  \
0  9.780440e+12               Suzanne Collins                       2008   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                       1997   
2  9.780320e+12               Stephenie Meyer                       2005   
3  9.780060e+12                    Harper Lee                       1960   
4  9.780740e+12           F. Scott Fitzgerald                       1925   

                             original_title  \
0                          The Hunger Games   
1  Harry Potter and the Philosopher's Stone   
2                                  Twilight   
3                     To Kill a Mockingbird   
4                          The Great Gatsby   

                         ...                         ratings_count  \
0                        ...                               4780653   
1                        ...                               4602479   
2                        ...                               3866839   
3                        ...                               3198671   
4                        ...                               2683664   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            4800065                    75867      75504     101676   
2            3916824                    95009     456191     436802   
3            3340896                    72586      60427     117415   
4            2773745                    51992      86236     197621   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     455024    1156318    3011543   
2     793319     875073    1355439   
3     446835    1001952    1714267   
4     606158     936012     947718   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1474154022s...  
2  https://images.gr-assets.com/books/1361039443s...  
3  https://images.gr-assets.com/books/1361975680s...  
4  https://images.gr-assets.com/books/1490528560s...  

[5 rows x 23 columns]

Let's work in pandas, and explore wildcards and regular expressions in SQL:

In [12]:
df = pd.read_sql('SELECT * FROM books WHERE original_title = "The Hunger Games"',con=connection)
df

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   

         isbn13          authors  original_publication_year    original_title  \
0  9.780440e+12  Suzanne Collins                       2008  The Hunger Games   

                         ...                         ratings_count  \
0                        ...                               4780653   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  

[1 rows x 23 columns]

In SQL, ```%``` means 1 or more wildcard characters; ```?``` means exactly one wildcard character.  So, if we wanted to look for titles that contain "Hunger Games", we could use:

In [13]:
df = pd.read_sql('SELECT * FROM books WHERE original_title LIKE "%Hunger Games%"',con=connection)
df

book_id  goodreads_book_id  best_book_id   work_id  books_count       isbn  \
0        1            2767052       2767052   2792775          272  439023483   
1      507            7938275       7938275  11349083           78  545265355   

         isbn13          authors  original_publication_year  \
0  9.780440e+12  Suzanne Collins                       2008   
1  9.780550e+12  Suzanne Collins                       2010   

             original_title  \
0          The Hunger Games   
1  The Hunger Games Box Set   

                         ...                         ratings_count  \
0                        ...                               4780653   
1                        ...                                129665   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1             159760                     6429       1015       2521   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1      13438      42932      99854   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1360094673m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1360094673s...  

[2 rows x 23 columns]

More powerfully, we could use regular expressions:

In [14]:
df = pd.read_sql('SELECT * FROM books WHERE original_title REGEXP ".*Hunger Games.*"',con=connection)
df

book_id  goodreads_book_id  best_book_id   work_id  books_count       isbn  \
0        1            2767052       2767052   2792775          272  439023483   
1      507            7938275       7938275  11349083           78  545265355   

         isbn13          authors  original_publication_year  \
0  9.780440e+12  Suzanne Collins                       2008   
1  9.780550e+12  Suzanne Collins                       2010   

             original_title  \
0          The Hunger Games   
1  The Hunger Games Box Set   

                         ...                         ratings_count  \
0                        ...                               4780653   
1                        ...                                129665   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1             159760                     6429       1015       2521   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1      13438      42932      99854   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1360094673m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1360094673s...  

[2 rows x 23 columns]

In [19]:
df = pd.read_sql('SELECT * FROM books WHERE original_publication_year IN (2008, 2010)',con=connection)
df.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count        isbn  \
0        1            2767052       2767052  2792775          272   439023483   
1       20            7260188       7260188  8812783          239   439023513   
2       56            1162543       1162543  2960529          183  031606792X   
3       73            1656001       1656001  3328799          161   316068047   
4       88            6442769       6442769  3364505          178  014241493X   

         isbn13          authors  original_publication_year    original_title  \
0  9.780440e+12  Suzanne Collins                       2008  The Hunger Games   
1  9.780440e+12  Suzanne Collins                       2010        Mockingjay   
2  9.780320e+12  Stephenie Meyer                       2008     Breaking Dawn   
3  9.780320e+12  Stephenie Meyer                       2008          The Host   
4  9.780140e+12       John Green                       2008       Paper Towns   

                         ...                         ratings_count  \
0                        ...                               4780653   
1                        ...                               1719760   
2                        ...                               1070245   
3                        ...                                749780   
4                        ...                                461311   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            1870748                    96274      30144     110498   
2            1107709                    44550     100994     114893   
3             777560                    39778      44215      62501   
4             655271                    42717      14105      47183   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     373060     618271     738775   
2     213402     267706     410714   
3     154906     227180     288758   
4     154479     223895     215609   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1358275419m...   
2  https://images.gr-assets.com/books/1361039438m...   
3  https://images.gr-assets.com/books/1318009171m...   
4  https://images.gr-assets.com/books/1349013610m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1358275419s...  
2  https://images.gr-assets.com/books/1361039438s...  
3  https://images.gr-assets.com/books/1318009171s...  
4  https://images.gr-assets.com/books/1349013610s...  

[5 rows x 23 columns]

Sometimes, it's nicer to wrap our SQL across multiple lines, so we can use docstrings ("""...""") to create our SQL.

Here's an example of a join using a WHERE statement:

In [20]:
df = pd.read_sql("""
                SELECT books.original_title, ratings.rating 
                FROM books,ratings 
                WHERE books.authors LIKE '%Rowling%' 
                AND books.book_id = ratings.book_id;
                """, con = connection)   

In [21]:
df.shape

(136631, 2)

In [22]:
df.head()

original_title  rating
0  Harry Potter and the Philosopher's Stone       5
1  Harry Potter and the Philosopher's Stone       4
2  Harry Potter and the Philosopher's Stone       4
3  Harry Potter and the Philosopher's Stone       5
4  Harry Potter and the Philosopher's Stone       3

In [23]:
df.sample(10)

original_title  rating
100765    Harry Potter and the Half-Blood Prince       2
87218       Harry Potter and the Deathly Hallows       5
102358    Harry Potter and the Half-Blood Prince       5
117328                      The Cuckoo's Calling       5
6327    Harry Potter and the Philosopher's Stone       5
60737    Harry Potter and the Chamber of Secrets       4
82486        Harry Potter and the Goblet of Fire       5
74603        Harry Potter and the Goblet of Fire       5
27812   Harry Potter and the Prisoner of Azkaban       4
133351                            Career of Evil       3

We can leverage the groupby function in pandas to do helpful calculations:

In [24]:
df.groupby(by='original_title')[['rating']].mean()

rating
original_title                                              
                                                    4.588123
Career of Evil                                      4.134006
Complete Harry Potter Boxed Set                     4.659530
Fantastic Beasts and Where to Find Them             3.775901
Fantastic Beasts and Where to Find Them: The Or...  4.384259
Harry Potter Boxed Set Books 1-4                    4.605357
Harry Potter Collection (Harry Potter, #1-6)        4.699571
Harry Potter and the Chamber of Secrets             4.229418
Harry Potter and the Cursed Child, Parts One an...  3.694737
Harry Potter and the Deathly Hallows                4.525941
Harry Potter and the Goblet of Fire                 4.430780
Harry Potter and the Half-Blood Prince              4.443339
Harry Potter and the Order of the Phoenix           4.358697
Harry Potter and the Philosopher's Stone            4.351350
Harry Potter and the Prisoner of Azkaban            4.418732
Harry, A History: The True Story of a Boy Wizar...  4.090909
Hogwarts: An Incomplete and Unreliable Guide        4.219858
Quidditch Through the Ages                          3.729560
Short Stories from Hogwarts of Heroism, Hardshi...  4.329268
Short Stories from Hogwarts of Power, Politics ...  4.173913
The Casual Vacancy                                  3.174991
The Cuckoo's Calling                                3.794902
The Silkworm                                        3.950251
The Tales of Beedle the Bard                        3.828964
Very Good Lives: The Fringe Benefits of Failure...  4.275510

Here's a schematic representation of the goodreads database:

![](assets/goodreadsEER.png)

In [25]:
df.sample(10)

original_title  rating
41938   Harry Potter and the Order of the Phoenix       3
21577    Harry Potter and the Philosopher's Stone       5
125164               The Tales of Beedle the Bard       3
74945         Harry Potter and the Goblet of Fire       5
116305                       The Cuckoo's Calling       4
59418     Harry Potter and the Chamber of Secrets       5
125413               The Tales of Beedle the Bard       3
86057        Harry Potter and the Deathly Hallows       5
40063   Harry Potter and the Order of the Phoenix       4
106934     Harry Potter and the Half-Blood Prince       5

### <font color="magenta">Q2: How many books in our dataset were originally published in 2016?</font>

In [31]:
df = pd.read_sql('SELECT * FROM books WHERE original_publication_year IN (2016)',con=connection)
df.head()

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0      279           29056083      29056083  48765776           95   
1      852           25899336      25899336  45424659           80   
2     1028           26247008      26247008  49997474           45   
3     1076           25781157      25781157  45630747           46   
4     1174           26893819      26893819  42856015           59   

         isbn        isbn13                                  authors  \
0   751565350  9.780750e+12  John Tiffany, Jack Thorne, J.K. Rowling   
1              0.000000e+00         Paul Kalanithi, Abraham Verghese   
2  1250069793  9.781250e+12                           Liane Moriarty   
3    62414216  9.780060e+12                  Cynthia D'Aprix Sweeney   
4  081299860X  9.780810e+12                               Emma Cline   

   original_publication_year  \
0                       2016   
1                       2016   
2                       2016   
3                       2016   
4                       2016   

                                      original_title  \
0  Harry Potter and the Cursed Child, Parts One a...   
1                            When Breath Becomes Air   
2                                 Truly Madly Guilty   
3                                           The Nest   
4                                          The Girls   

                         ...                         ratings_count  \
0                        ...                                270603   
1                        ...                                116771   
2                        ...                                 90477   
3                        ...                                 79762   
4                        ...                                 76761   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0             397773                    53365      15828      35842   
1             165673                    15933       2913       4244   
2             116339                    11983       2752      10961   
3              96364                    10214       2615       9810   
4              94812                    11573       2905      11033   

   ratings_3  ratings_4  ratings_5  \
0      96395     133156     116552   
1      18562      51535      88419   
2      41045      42953      18628   
3      36492      36673      10774   
4      32122      34599      14153   

                                           image_url  \
0  https://images.gr-assets.com/books/1470082995m...   
1  https://images.gr-assets.com/books/1492677644m...   
2  https://images.gr-assets.com/books/1491249206m...   
3  https://images.gr-assets.com/books/1461354827m...   
4  https://images.gr-assets.com/books/1492065338m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1470082995s...  
1  https://images.gr-assets.com/books/1492677644s...  
2  https://images.gr-assets.com/books/1491249206s...  
3  https://images.gr-assets.com/books/1461354827s...  
4  https://images.gr-assets.com/books/1492065338s...  

[5 rows x 23 columns]

198 books in dataset published in 2016

### <font color="magenta">Q3: How many books were published in each year of original publication in our dataset?</font>


In [29]:
df = pd.read_sql("""
        select original_publication_year as year, count(*) as count
        from books group by original_publication_year
        """,con=connection)
df

year  count
0   -1750      1
1    -762      1
2    -750      2
3    -720      1
4    -560      1
5    -500      3
6    -476      1
7    -458      2
8    -441      1
9    -440      1
10   -431      1
11   -430      1
12   -411      2
13   -401      1
14   -400      2
15   -390      1
16   -385      2
17   -380      1
18   -350      2
19   -335      1
20   -330      1
21   -300      1
22    -17      1
23      0     21
24      8      1
25    119      1
26    180      1
27    397      1
28    609      1
29    800      1
..    ...    ...
264  1988     89
265  1989     87
266  1990     99
267  1991     98
268  1992    112
269  1993    115
270  1994    121
271  1995    126
272  1996    163
273  1997    168
274  1998    172
275  1999    186
276  2000    209
277  2001    226
278  2002    225
279  2003    288
280  2004    307
281  2005    326
282  2006    362
283  2007    363
284  2008    383
285  2009    432
286  2010    473
287  2011    556
288  2012    568
289  2013    518
290  2014    437
291  2015    306
292  2016    198
293  2017     11

[294 rows x 2 columns]

### <font color="magenta">Q4: Do you notice anything interesting or strange about your results from the previous question?  Investigate at least one of these anomalies using either an SQL query or pandas filtering. </font>

Insert your answer here

There are negative year for some books. These books are represented as B.C. dates. I searched for the year using SQL and found the book related to the negative year. Then I googled the book and found out the book was written in B.C.

### <font color="magenta">Q5: Create a dataframe that contains a list of all book titles from 2017 as well as the tag_names associated with each book.</font>

In [33]:
df = pd.read_sql("""SELECT books.title, tags.tag_name
FROM books
LEFT JOIN books_tags on books_tags.goodreads_book_id = books.goodreads_book_id
LEFT JOIN tags on tags.tag_id = books_tags.tag_id
WHERE books.original_publication_year=2017""", con=connection)

In [34]:
df2=df.groupby(by='title').sum()

In [35]:
df['tags']=df2['tag_name'].str.split('\r')

In [36]:
df2.head(2)

tag_name
title                                                                                                
A Court of Wings and Ruin (A Court of Thorns an...  to-read\rcurrently-reading\rfantasy\rfavorites...
Behind Her Eyes                                     to-read\rcurrently-reading\rthriller\rmystery\...

# END OF NOTEBOOK: REMEMBER TO STRIP IDETIFYING INFORMATION FROM TOP OF NOTEBOOK!
Please submit both ipynb and html formats of this notebook to Canvas.